### ISSUE TITLE
Creating a map from a .fz file causes an error 

### ISSUE THREAD

**Issue by samaloney** [sunpy/sunpy#6655]
### Describe the bug
Originally raised by nawinnga in the sunpy chat room.

Opening and GONG H-alpha fits obtained from Fido causes an error, it tires to read it as an ana file.

### To Reproduce
```
import sunpy
sunpy.map.Map('ftp://gong2.nso.edu/HA/haf/202203/20220318/20220318000050Bh.fits.fz')

ck_synch_hd: error: file does not have the F0 synch pattern (found 0x504d4953 instead)
ana_fzread: error: ck_sync_hd error!
Traceback (most recent call last):
  File "/Users/shane/.virtualenvs/temp/lib/python3.7/site-packages/sunpy/map/map_factory.py", line 161, in _read_file
    pairs = read_file(os.fspath(fname), **kwargs)
  File "/Users/shane/.virtualenvs/temp/lib/python3.7/site-packages/sunpy/io/file_tools.py", line 89, in read_file
    return _readers[readername].read(filepath, **kwargs)
  File "/Users/shane/.virtualenvs/temp/lib/python3.7/site-packages/sunpy/io/ana.py", line 53, in read
    data = _pyana.fzread(filename, debug)
ValueError: In pyana_fzread: could not read ana file, data returned is NULL.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/shane/.virtualenvs/temp/lib/python3.7/site-packages/sunpy/map/map_factory.py", line 318, in __call__
    data_header_pairs = self._parse_args(*args, silence_errors=silence_errors, **kwargs)
  File "/Users/shane/.virtualenvs/temp/lib/python3.7/site-packages/sunpy/map/map_factory.py", line 244, in _parse_args
    data_header_pairs += self._parse_arg(arg, **kwargs)
  File "/Users/shane/.virtualenvs/temp/lib/python3.7/site-packages/sunpy/util/functools.py", line 18, in wrapper
    return dispatcher.dispatch(args[1].__class__)(*args, **kwargs)
  File "/Users/shane/.virtualenvs/temp/lib/python3.7/site-packages/sunpy/map/map_factory.py", line 285, in _parse_url
    pairs = self._read_file(path, **kwargs)
  File "/Users/shane/.virtualenvs/temp/lib/python3.7/site-packages/sunpy/map/map_factory.py", line 164, in _read_file
    raise IOError(msg) from e
OSError: Failed to read /Users/shane/Library/Caches/sunpy/20220318000050Bh.fits.fz.
```

### System Details

OS: Mac OS 10.16 Arch: 64bit, (i386)
python3.9
sunpy: 4.1.0
astropy: 5.1.1
numpy: 1.23.5

### Installation method

pip

**Comment by markandey1414** [comment]
I directly tried importing the map, it still raises the issue that there's no C extension for ANA.


`
import sunpy.map
img = sunpy.map.Map('ftp://gong2.nso.edu/HA/haf/202203/20220318/20220318000050Bh.fits.fz')



Traceback (most recent call last):
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 161, in _read_file
    pairs = read_file(os.fspath(fname), **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\io\file_tools.py", line 87, in read_file
    return _readers[readername].read(filepath, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\io\ana.py", line 51, in read
    raise ImportError("C extension for ANA is missing, please rebuild.")
ImportError: C extension for ANA is missing, please rebuild.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\marka\PycharmProjects\openAstronomy\main.py", line 2, in <module>
    aia = sunpy.map.Map('ftp://gong2.nso.edu/HA/haf/202203/20220318/20220318000050Bh.fits.fz')
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 318, in __call__
    data_header_pairs = self._parse_args(*args, silence_errors=silence_errors, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 244, in _parse_args
    data_header_pairs += self._parse_arg(arg, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\util\functools.py", line 18, in wrapper
    return dispatcher.dispatch(args[1].__class__)(*args, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 285, in _parse_url
    pairs = self._read_file(path, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 164, in _read_file
    raise IOError(msg) from e
OSError: Failed to read C:\Users\marka\AppData\Local\sunpy\sunpy\Cache\20220318000050Bh.fits.fz
`

**Comment by markandey1414** [comment]
`import sunpy.map
img = sunpy.map.Map('ftp://gong2.nso.edu/HA/haf/202203/20220318/20220318000050Bh.fits.fz')

Traceback (most recent call last):
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 161, in _read_file
    pairs = read_file(os.fspath(fname), **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\io\file_tools.py", line 87, in read_file
    return _readers[readername].read(filepath, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\io\ana.py", line 51, in read
    raise ImportError("C extension for ANA is missing, please rebuild.")
ImportError: C extension for ANA is missing, please rebuild.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\marka\PycharmProjects\openAstronomy\main.py", line 2, in <module>
    aia = sunpy.map.Map('ftp://gong2.nso.edu/HA/haf/202203/20220318/20220318000050Bh.fits.fz')
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 318, in __call__
    data_header_pairs = self._parse_args(*args, silence_errors=silence_errors, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 244, in _parse_args
    data_header_pairs += self._parse_arg(arg, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\util\functools.py", line 18, in wrapper
    return dispatcher.dispatch(args[1].__class__)(*args, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 285, in _parse_url
    pairs = self._read_file(path, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 164, in _read_file
    raise IOError(msg) from e
OSError: Failed to read C:\Users\marka\AppData\Local\sunpy\sunpy\Cache\20220318000050Bh.fits.fz
`

**Comment by markandey1414** [comment]
It raises the issue that the C extension for the ANA file is missing


`
import sunpy.map
img = sunpy.map.Map('ftp://gong2.nso.edu/HA/haf/202203/20220318/20220318000050Bh.fits.fz')



Traceback (most recent call last):
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 161, in _read_file
    pairs = read_file(os.fspath(fname), **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\io\file_tools.py", line 87, in read_file
    return _readers[readername].read(filepath, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\io\ana.py", line 51, in read
    raise ImportError("C extension for ANA is missing, please rebuild.")
ImportError: C extension for ANA is missing, please rebuild.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\marka\PycharmProjects\openAstronomy\main.py", line 2, in <module>
    aia = sunpy.map.Map('ftp://gong2.nso.edu/HA/haf/202203/20220318/20220318000050Bh.fits.fz')
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 318, in __call__
    data_header_pairs = self._parse_args(*args, silence_errors=silence_errors, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 244, in _parse_args
    data_header_pairs += self._parse_arg(arg, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\util\functools.py", line 18, in wrapper
    return dispatcher.dispatch(args[1].__class__)(*args, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 285, in _parse_url
    pairs = self._read_file(path, **kwargs)
  File "C:\Users\marka\PycharmProjects\openAstronomy\venv\lib\site-packages\sunpy\map\map_factory.py", line 164, in _read_file
    raise IOError(msg) from e
OSError: Failed to read C:\Users\marka\AppData\Local\sunpy\sunpy\Cache\20220318000050Bh.fits.fz`

**Comment by Cadair** [comment]
The two tracebacks are the same thing, which is that our automatic file detection code prefers to dispatch on file extension and not magic bytes:

https://github.com/sunpy/sunpy/blob/6015212d2c2b43ff804ea7644eaaab4da3da43e9/sunpy/io/file_tools.py#L84-L91

This means we treat `.fz` files as ANA files, which either causes the loader to error because it's not a ANA file or the loader to fail to load because it doesn't work on windows.

My initial reaction would be that we should probably try and use magic bytes first as a file detection tool as you are much less likely to get false detections, and then fall back to file extension afterwards. The only downside of this is you need to open the file.

**Comment by Cadair** [comment]
On a slightly different note while I am staring at this code, this is just mad to have inside the `detect_filetype`:

https://github.com/sunpy/sunpy/blob/6015212d2c2b43ff804ea7644eaaab4da3da43e9/sunpy/io/file_tools.py#L217-L224

We should be able to detect gzip, decompress some bytes and then detect FITS.

**Comment by nabobalis** [comment]
I think this might be solved by https://github.com/sunpy/sunpy/pull/6693

**Comment by samaloney** [comment]
I think #6693 may help but the `.fz` extension will still cause issues

**Comment by dstansby** [comment]
Given I'm not sure this ever worked, I've swapped the labels from Bug to Feature Request

**Comment by ayshih** [comment]
> In another twist to this story VSO client is doing interesting things to the file name in this case which is returned from the VSO as `20240209120002Uh.fits.fz` but the `slugify` function turns this into `20240209120002uh_fits.fz`. This doesn't really make any difference to this specific issue but I guess it would be better to not alter a valid filename?

For posterity, this issue was separately reported in #7450 and fixed in #7453

**Comment by nabobalis** [comment]
I think we fixed this. 

Can someone check?

**Comment by Cadair** [comment]
I just did and we have not.

**Comment by samaloney** [comment]
Hopefully now fixed in #7788 - we were detecting the filetype but then not using it 🤦

**Comment by samaloney** [comment]
Ok just did this on master and file opened as expected 

```
m = Map("http://gong2.nso.edu/HA/haf/202203/20220318/20220318000050Bh.fits.fz")
```
